File: tool/check_l10n_keys.mjs
import fs from 'fs';

const text = fs.readFileSync('lib/localization/app_localizations.dart', 'utf8');

function keysFor(lang) {
  const marker = `'${lang}': {`;
  const start = text.indexOf(marker);
  if (start < 0) return new Set();
  let i = start + marker.length;
  let depth = 1;
  while (i < text.length && depth > 0) {
    if (text[i] === '{') depth += 1;
    else if (text[i] === '}') depth -= 1;
    i += 1;
  }
  const block = text.slice(start + marker.length, i - 1);
  return new Set([...block.matchAll(/'([^']+)':/g)].map((m) => m[1]));
}

const en = keysFor('en');
const hi = keysFor('hi');
const mr = keysFor('mr');
const missingHi = [...en].filter((k) => !hi.has(k)).sort();
const missingMr = [...en].filter((k) => !mr.has(k)).sort();

console.log(`EN ${en.size} | HI ${hi.size} | MR ${mr.size}`);
console.log(`Missing in HI: ${missingHi.length}`);
console.log(missingHi.join('\n'));
console.log(`Missing in MR: ${missingMr.length}`);
console.log(missingMr.join('\n'));

File: tool/check_l10n_keys.py
import re
from pathlib import Path

text = Path('lib/localization/app_localizations.dart').read_text(encoding='utf-8')

def keys_for(lang):
    m = re.search(rf"'{lang}':\s*\{{", text)
    if not m:
        return set()
    start = m.end()
    depth = 1
    i = start
    while i < len(text) and depth:
        if text[i] == '{':
            depth += 1
        elif text[i] == '}':
            depth -= 1
        i += 1
    block = text[start:i - 1]
    return set(re.findall(r"'([^']+)':", block))

en, hi, mr = keys_for('en'), keys_for('hi'), keys_for('mr')
print('EN', len(en), 'HI', len(hi), 'MR', len(mr))
print('Missing in HI', len(en - hi))
for k in sorted(en - hi):
    print('  HI missing:', k)
print('Missing in MR', len(en - mr))
for k in sorted(en - mr):
    print('  MR missing:', k)
